### DISCUSSION TITLE
Why does logic in `processed_file_names` make `InMemoryDataset` forget about `processed_paths`?

### DISCUSSION THREAD

**Discussion by NiklasKappel** [pyg-team/pytorch_geometric#8811] [Q&A] [answered]
Consider this minimal (not) working example:
```python
# file my_dataset.py

from torch_geometric.data import Data, InMemoryDataset


class MyOwnDataset(InMemoryDataset):
    def __init__(
        self,
        root,
        transform=None,
        pre_transform=None,
        pre_filter=None,
        remove_outliers=True,
    ):
        super().__init__(root, transform, pre_transform, pre_filter)
        self.remove_outliers = remove_outliers
        self.load(self.processed_paths[0])

    @property
    def raw_file_names(self):
        return ["data.txt"]

    @property
    def processed_file_names(self):
        if self.remove_outliers:
            return ["data_wo_outliers.pt"]
        return ["data.pt"]

    def process(self):
        data_list = []

        # Read data file
        data_list.append(Data())

        if self.remove_outliers:
            # Remove outliers
            pass

        self.save(data_list, self.processed_paths[0])


if __name__ == "__main__":
    dataset = MyOwnDataset(root="data")
    print(dataset[0])
```

Running `python my_dataset.py` crashes with:
```AttributeError: 'MyOwnDataset' object has no attribute 'processed_paths'. Did you mean: 'processed_dir'?```

If I comment out only the if condition in `processed_file_names`, the code works (assuming there is a file at `data/raw/data.txt`). I don't get how the presence of an if condition one method (`processed_file_names`) can determine the existence of a completely different method (`processed_paths`) in the parent class. I am not sure if this is a bug with PyG, or if I am doing something wrong here.

Edit: I can also make it work by keeping the if condition, but swapping the first to lines of `__init__`, that is the super call and the `remove_outliers` assignment. I have don't get why though, which makes me feel pretty uncomfortable with my solution.

**Comment by rusty1s** [comment, marked as answer]
Thanks for the issue. Python will output this error because there is an error raised when accessing `processed_file_names`. In particular, because `self.remove_outliers` is not yet defined since `processed_file_names` is called in `super().__init__()`.

This fixes it:
```python
        self.remove_outliers = remove_outliers
        super().__init__(root, transform, pre_transform, pre_filter)
        self.load(self.processed_paths[0])
